Adds support for building libsodium on Windows.

AndreLouisCaron opened the pull request:
Libsodium compiles fine on Windows, even with VS2008 (for Python 2.7).  However, it requires a few tricks:
- VS2008 doesn't provide `<stdint.h>`, so we need to mock it (like libzmq already does);
- VS2008 doesn't support C99 `static inline` functions, but removing the `inline` when compiling it for PyZMQ seems to work fine;
- before version 1.0.1, libsodium uses the same `DLL_EXPORT` switch as libzmq, which prevented from compiling libsodium as a dependency of libzmq (since version 1.0.1, libsodium changed this macro to `SODIUM_DLL_EXPORT` and no longer conflicts with PyZMQ).

AndreLouisCaron commented:
I've only tested this against Python 2.7.9 on Windows.  I assume this needs testing on other supported Python versions!

minrk commented on setup.py:
```diff
@@ -304,6 +304,8 @@ def init_settings_from_config(self):
             'devices',
         )]
+        if sys.platform.startswith('win'):
+            settings['include_dirs'].insert(0, 'patch')
```
it seems like this should just affect libsodium, yes?

AndreLouisCaron replied:
When I remove it, I get this error:

    monitoredqueue.c
    bundled\zeromq\include\zmq.h(78) : fatal error C1083: Cannot open include file: 'stdint.h': No such file or directory

The `stdint.hpp` it provides is used only internally.  I guess `libzmq` also depends on `<stdint.h>` being available to client applications...

minrk replied:
>  I guess libzmq also depends on <stdint.h> being available to client applications...

That seems improbable, since it builds successfully without it in master.

AndreLouisCaron replied:
I just checked and it seems `zmq.h` in 4.0.x and 4.1.x provide their own definitions of integer types but master does not.  See:
- [zmq.h line 73 in zeromq/libzmq](https://github.com/zeromq/libzmq/blob/master/include/zmq.h#L73) -- no integer types,
- [zmq.h line 73 in zeromq/zeromq4-x](https://github.com/zeromq/zeromq4-x/blob/master/include/zmq.h#L73) -- integer types; and
- [zmq.h line 73 in zeromq/zeromq4-1](https://github.com/zeromq/zeromq4-1/blob/master/include/zmq.h#L73) -- integer types.

I'm guessing master is "wrong" here, but the stable fork's definitions are kind of strange (is it even legal C to use preprocessor directives to see if a `typedef` has been defined?)

minrk commented on setup.py:
```diff
@@ -510,7 +525,8 @@ def bundle_libzmq_extension(self):
 
-            libzmq.libraries.extend(['rpcrt4', 'ws2_32', 'advapi32'])
+            libzmq.libraries.extend(['rpcrt4', 'ws2_32', 'advapi32', 'libsodium'])
+            libzmq.library_dirs.append('build/temp.win32-2.7/Release/buildutils')
```
This is definitely not going to work anywhere but 2.7.

AndreLouisCaron replied:
Fixed by copying similar code for `libzmq` in the top-level `bundled_settings` function.

minrk commented on patch/stdint.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef _stdint_h__
```
This should be in `buildutils/include_win32`

AndreLouisCaron replied:
Fixed.

minrk commented:
Thanks, this is super helpful. I think it's going to need a bit of help on Python 3, and I think some of the additions may not be necessary, but we can work it out.

AndreLouisCaron commented:
Cool!  I'll start by addressing the issues you raised and the. I'll try it on Python 3 and update it he pull request!

minrk commented:
Terrific, thanks.

AndreLouisCaron commented:
Updated the pull request.  Tested on Python 2.7.9 and 3.4.3.  I think I addressed all issues, let me know if there's anything else.

I'm not sure how to test all build variants (debug/release/64-bit).  Do you have a procedure somewhere so that I can try them out or are you going to test it yourself?

minrk commented:
You can look at the script in `tools/release_windows.bat`, which does the Windows binary builds. It could be a starting point for testing.

minrk commented on setup.py:
```diff
@@ -499,6 +513,8 @@ def bundle_libzmq_extension(self):
             libzmq.define_macros.append(('FD_SETSIZE', 1024))
             libzmq.define_macros.append(('DLL_EXPORT', 1))
+            libzmq.define_macros.append(('_CRT_SECURE_NO_WARNINGS', 1))
+            libzmq.define_macros.append(('ZMQ_USE_SELECT', 1))
```
why have you changed the value for ZMQ_USE_SELECT?

AndreLouisCaron replied:
When I don't set this, I get the following error:

    c:\users\acaron\desktop\develop\pyzmq\bundled\zeromq\src\poller.hpp(42) : fatal error C1189: #error :  None of the ZMQ_USE_* macros defined

I get this when building from master with out my changes too.  The one thing I'm doing differently is building against the latest ZeroMQ master, but I can't understand why that's creating the error.  Any ideas?

minrk replied:
> The one thing I'm doing differently is building against the latest ZeroMQ master, but I can't understand why that's creating the error.

That could easily be the reason. Perhaps this changed in libzmq master.

AndreLouisCaron replied:
I haven't had time to test this yet, but since `ZMQ_USE_SELECT` is the only poller choice available on Windows, I guess it doesn't hurt to provide this definition.

minrk commented:
Thanks, this is looking good. I'll try to find some time to test it (I may not be able to for the next two weeks, due to conferences and travel).

AndreLouisCaron commented:
Hi, did you have time to try this out?

Thanks!

minrk commented:
@AndreLouisCaron sorry, not yet. I'm in the middle of a big project on IPython that's got some time pressure. I'll try to get to this soon. Thanks for your patience.

minrk commented:
@AndreLouisCaron I was able to do a bit more review and testing, and it all looks and works great. Thanks!
